radicle-cli/src/commands/merge.rs

```diff
@@ -0,0 +1,385 @@
+        //
+        // Let's check if there are potential merge conflicts.
+        let our_commit = head.peel_to_commit()?;
+        let their_commit = repo.find_commit(revision.oid.into())?;
+
+        let index = repo
+            .merge_commits(&our_commit, &their_commit, None)
+            .context("failed to perform merge analysis")?;
+
+        if index.has_conflicts() {
+            return Err(Error::WithHint {
+                err: anyhow!("patch conflicts with {}", branch),
+                hint: "Patch must be rebased before it can be merged.",
+            }
+            .into());
+        }
+        MergeStyle::Commit
+    } else if merge.is_up_to_date() {
+        term::info!(
+            "✓ Patch {} is already part of {}",
+            term::format::tertiary(patch_id),
+            term::format::highlight(branch)
+        );
+
+        return Ok(());
+    } else if merge.is_unborn() {
+        anyhow::bail!("HEAD does not point to a valid commit");
+    } else {
+        anyhow::bail!(
+            "no merge is possible between {} and {}",
+            head_oid,
+            revision.oid
+        );
+    };
+
+    let merge_style_pretty = match merge_style {
+        MergeStyle::FastForward => term::format::style(merge_style.to_string())
+            .dim()
+            .italic()
+            .to_string(),
+        MergeStyle::Commit => term::format::style(merge_style.to_string())
+            .yellow()
+            .italic()
+            .to_string(),
+    };
+
+    term::info!(
+        "{} {} {} ({}) by {} into {} ({}) via {}...",
+        term::format::bold("Merging"),
+        term::format::tertiary(term::format::cob(&patch_id)),
+        term::format::dim(format!("R{}", revision_id)),
+        term::format::secondary(term::format::oid(revision.oid)),
+        term::format::tertiary(patch.author.id),
+        term::format::highlight(branch),
+        term::format::secondary(term::format::oid(head_oid)),
+        merge_style_pretty
+    );
+
+    if options.interactive && !term::confirm("Confirm?") {
+        anyhow::bail!("merge aborted by user");
+    }
+
+    //
+    // Perform merge
+    //
+    match merge_style {
+        MergeStyle::Commit => {
+            merge_commit(&repo, patch_id, &patch_commit, &patch, cobs.public_key())?;
+        }
+        MergeStyle::FastForward => {
+            fast_forward(&repo, &revision.oid)?;
+        }
+    }
+
+    term::success!(
+        "Updated {} {} -> {} via {}",
+        term::format::highlight(branch),
+        term::format::secondary(term::format::oid(head_oid)),
+        term::format::secondary(term::format::oid(revision.oid)),
+        merge_style_pretty
+    );
+
+    //
+    // Update patch COB
+    //
+    // TODO: Don't allow merging the same revision twice?
+    patches.merge(&patch_id, revision_id, head_oid.into(), &signer)?;
+
+    term::success!(
+        "Patch state updated, use {} to publish",
+        term::format::secondary("`rad push`")
+    );
+
+    Ok(())
```

Comment:
Could be split into functions to improve clarity

Reply:
Let's explore this idea in a pairing session at some point; I don't think it's an issue in the short term.